pkg/packagekit/updates.jsx

```diff
@@ -402,6 +436,108 @@ const UpdatesList = ({ updates }) => {
     );
 };
 
+class RestartServices extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            selected: props.systemctl,
+            showDialog: false,
+        };
+
+        this.dialogErrorSet = this.dialogErrorSet.bind(this);
+        this.dialogErrorDismiss = this.dialogErrorDismiss.bind(this);
+        this.onValueChanged = this.onValueChanged.bind(this);
+        this.restart = this.restart.bind(this);
+        this.open = this.open.bind(this);
+        this.close = this.close.bind(this);
+        this.onRowToggle = this.onRowToggle.bind(this);
+    }
+
+    dialogErrorSet(text, detail) {
+        this.setState({ dialogError: text, dialogErrorDetail: detail });
+    }
+
+    dialogErrorDismiss() {
+        this.setState({ dialogError: undefined });
+    }
+
+    open() {
+        this.setState({ showDialog: true });
+    }
+
+    close() {
+        this.setState({ showDialog: false });
+    }
+
+    onValueChanged(key, value) {
+        this.setState({ [key]: value });
+    }
+
+    onRowToggle(service) {
+        const index = this.state.selected.indexOf(service);
+        const selected = [...this.state.selected];
+
+        if (index > -1)
+            delete selected[index];
+        else
+            selected.push(service);
+
+        this.setState({ selected });
+    }
+
+    restart() {
+        const services = this.state.selected;
+        Promise.all(services.map(service => cockpit.dbus("org.freedesktop.systemd1", { superuser: "try" }).call("/org/freedesktop/systemd1", "org.freedesktop.systemd1.Manager", "RestartUnit", [service + ".service", "fail"]))) // TODO be smarter than just concatenating ".service"
+                .then(() => {
+                    this.props.callTracer(); // TODO clean systemctl
+                    this.close();
+                })
+                .catch(ex => console.warn("RestartUnit: failed to restart a unit:", ex.message));
+    }
+
+    render() {
+        const rows = this.props.systemctl.map(service => { return { id: service, name: service, selected: this.state.selected.indexOf(service) > -1 } });
+
+        const body = (<>
+            <SelectableListing rows={rows} onRowToggle={this.onRowToggle} idPrefix="restart-services" />
+            {_("Additionally the following services need to be restarted manually: ") + this.props.manual.join(", ")}
```

Comment:
If this happens, let's offer a "Reboot" button, as that's often preferable than manually fiddling with everything